unittest/gunit/keyring_vault/vault_keyring-api-t.cc

```diff
@@ -0,0 +1,340 @@
+#include <my_global.h>
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+#include <sql_plugin_ref.h>
+#include <fstream>
+#include "vault_keyring.cc"
+#include "keyring_impl.cc"
+#include "mock_logger.h"
+
+namespace keyring__api_unittest
+{
+  using ::testing::StrEq;
+  using namespace keyring;
+
+  class Keyring_vault_api_test : public ::testing::Test
+  {
+  public:
+    Keyring_vault_api_test()
+    {}
+    ~Keyring_vault_api_test()
+    {
+      delete[] plugin_name;
+      delete[] keyring_filename;
+    }
+
+    static std::string correct_token;
+
+  protected:
+    virtual void SetUp()
+    {
+      plugin_name= new char[strlen("FakeKeyring")+1];
+      strcpy(plugin_name, "FakeKeyring");
+      keyring_filename= new char[strlen("./keyring_vault.conf")+1];
+      strcpy(keyring_filename, "./keyring_vault.conf");
+
+      plugin_info.name.str= plugin_name;
+      plugin_info.name.length= strlen(plugin_name);
+      keyring_vault_config_file= keyring_filename;
+
+      keyring_init_with_mock_logger();
+
+      key_memory_KEYRING= PSI_NOT_INSTRUMENTED;
+      key_LOCK_keyring= PSI_NOT_INSTRUMENTED;
+      sample_key_data= "Robi";
+    }
+    virtual void TearDown()
+    {
+      keyring_deinit_with_mock_logger();
+    }
+  protected:
+    void keyring_init_with_mock_logger();
+    void keyring_deinit_with_mock_logger();
+
+    std::string sample_key_data;
+    char *plugin_name;
+    char *keyring_filename;
+    st_plugin_int plugin_info; //for Logger initialization
+  };
+
+  std::string Keyring_vault_api_test::correct_token;
+
+  void Keyring_vault_api_test::keyring_init_with_mock_logger()
+  {
+    ASSERT_FALSE(keyring_vault_init(&plugin_info));
+    //use MockLogger instead of Logger
+    logger.reset(new Mock_logger());
+  }
+
+  void Keyring_vault_api_test::keyring_deinit_with_mock_logger()
+  {
+    keyring_vault_deinit(NULL);
+  }
+
+  TEST_F(Keyring_vault_api_test, StoreFetchRemove)
+  {
+    EXPECT_FALSE(mysql_key_store("Robert_key", "AES", "Robert", sample_key_data.c_str(),
+                              sample_key_data.length()));
+    char *key_type;
+    size_t key_len;
+    void *key;
+    EXPECT_FALSE(mysql_key_fetch("Robert_key", &key_type, "Robert", &key,
+                              &key_len));
+    EXPECT_STREQ("AES", key_type);
+    EXPECT_EQ(key_len, sample_key_data.length());
+    ASSERT_TRUE(memcmp((char *)key, sample_key_data.c_str(), key_len) == 0);
+    my_free(key_type);
+    key_type= NULL;
+    my_free(key);
+    key= NULL;
+    EXPECT_FALSE(mysql_key_remove("Robert_key", "Robert"));
+    //make sure the key was removed - fetch it
+    EXPECT_FALSE(mysql_key_fetch("Robert_key", &key_type, "Robert", &key,
+                                 &key_len));
+    ASSERT_TRUE(key == NULL);
+  }
+
+  TEST_F(Keyring_vault_api_test, CheckIfInmemoryKeyIsNOTXORed)
```

Comment:
In keyring_file before a key data was stored in memory it was first XORed with predefined string. Later after key's data was fetched from in-memory cache (keys_container) it was XORed once again. Obfuscating data did not make a lot of sense. Now it is not XORed, thus to make sure this is the case - I first use keyring API to store a key, then extract the key directly from the cache to see if it is the same as the one I put there with keyring API. In keyring_file I had to first XOR it before I could check for equality.